packages/dev/src/lib/env.ts

```diff
@@ -0,0 +1,427 @@
+import type { NetlifyAPI } from '@netlify/api'
+import type { EnvironmentVariables } from '@netlify/types'
+
+export type EnvironmentVariableSource = 'account' | 'addons' | 'configFile' | 'general' | 'internal' | 'ui'
+
+/**
+ * Supported values for the user-provided env `context` option.
+ * These all match possible `context` values returned by the Envelope API.
+ * Note that a user may also specify a branch name with the special `branch:my-branch-name` format.
+ */
+export const SUPPORTED_CONTEXTS = ['all', 'production', 'deploy-preview', 'branch-deploy', 'dev'] as const
+
+/**
+ * Additional aliases for the user-provided env `context` option.
+ */
+const SUPPORTED_CONTEXT_ALIASES = {
+  dp: 'deploy-preview',
+  prod: 'production',
+}
+/**
+ * Supported values for the user-provided env `scope` option.
+ * These exactly match possible `scope` values returned by the Envelope API.
+ * Note that `any` is also supported.
+ */
+export const ALL_ENVELOPE_SCOPES = ['builds', 'functions', 'runtime', 'post_processing'] as const
+
+export type EnvironmentVariable = {
+  sources: EnvironmentVariableSource[]
+  value: string
+  scopes?: (typeof ALL_ENVELOPE_SCOPES)[]
+}
+
+type InjectedEnvironmentVariable = {
+  isInternal: boolean
+  originalValue: string | undefined
+  overriddenSources: string[]
+  usedSource: string
+}
+
+interface InjectEnvironmentVariablesOptions {
+  accountSlug?: string
+  baseVariables: Record<string, EnvironmentVariable>
+  envAPI: EnvironmentVariables
+  netlifyAPI?: NetlifyAPI
+  siteID?: string
+}
+
+export const injectEnvVariables = async ({
+  accountSlug,
+  baseVariables = {},
+  envAPI,
+  netlifyAPI,
+  siteID,
+}: InjectEnvironmentVariablesOptions) => {
+  const results: Record<string, InjectedEnvironmentVariable> = {}
+
+  let variables = baseVariables
+
+  if (netlifyAPI && accountSlug && siteID) {
```

Comment:
I notice there are multiple layers of nil handling for siteID and accountSlug here and in the callers. I'm finding hard to tell what's intentional behaviour, what's handling error cases, what's working around incomplete types, etc. Could we clarify this somehow? maybe comments at the very least? (is the double handling necessary at all?)

Reply:
https://github.com/netlify/primitives/pull/197/commits/d5b9a4fe284392ec216a44c452e4fa2850d6c344

Let me know if this address it.